File: lib/utils/polygonContainment.ts
import {
  doSegmentsIntersect,
  getSegmentIntersection,
  pointToSegmentDistance,
} from "@tscircuit/math-utils"
import { minimumDistanceBetweenSegments } from "lib/utils/minimumDistanceBetweenSegments"

export type PolygonPoint = { x: number; y: number }

const EPSILON = 1e-6

const isPointOnSegment = (
  point: PolygonPoint,
  start: PolygonPoint,
  end: PolygonPoint,
) => pointToSegmentDistance(point, start, end) <= EPSILON

const arePointsClose = (a: PolygonPoint, b: PolygonPoint) =>
  Math.abs(a.x - b.x) <= EPSILON && Math.abs(a.y - b.y) <= EPSILON

export const isPointInOrOnPolygon = (
  point: PolygonPoint,
  polygon: PolygonPoint[],
) => {
  if (!polygon || polygon.length < 3) {
    return false
  }

  // Treat points on the boundary as inside the polygon
  for (let i = 0; i < polygon.length; i++) {
    const start = polygon[i]
    const end = polygon[(i + 1) % polygon.length]
    if (isPointOnSegment(point, start, end)) {
      return true
    }
  }

  let inside = false

  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const pi = polygon[i]
    const pj = polygon[j]

    const intersect =
      pi.y > point.y !== pj.y > point.y &&
      point.x < ((pj.x - pi.x) * (point.y - pi.y)) / (pj.y - pi.y) + pi.x

    if (intersect) {
      inside = !inside
    }
  }

  return inside
}

export const doesSegmentCrossPolygonBoundary = ({
  start,
  end,
  polygon,
  margin = 0.2,
}: {
  start: PolygonPoint
  end: PolygonPoint
  polygon: PolygonPoint[]
  margin?: number
}) => {
  if (!polygon || polygon.length < 3) {
    return false
  }

  const startInside = isPointInOrOnPolygon(start, polygon)
  const endInside = isPointInOrOnPolygon(end, polygon)

  if (!startInside || !endInside) {
    return true
  }

  for (let i = 0; i < polygon.length; i++) {
    const edgeStart = polygon[i]
    const edgeEnd = polygon[(i + 1) % polygon.length]

    const startOnEdge = isPointOnSegment(start, edgeStart, edgeEnd)
    const endOnEdge = isPointOnSegment(end, edgeStart, edgeEnd)

    if (startOnEdge && endOnEdge) {
      // Moving along the outline edge is permitted
      continue
    }

    if (!doSegmentsIntersect(start, end, edgeStart, edgeEnd)) {
      if (!startOnEdge && !endOnEdge) {
        const distanceToEdge = minimumDistanceBetweenSegments(
          start,
          end,
          edgeStart,
          edgeEnd,
        )

        if (distanceToEdge < margin - EPSILON) {
          return true
        }
      }

      continue
    }

    const intersection = getSegmentIntersection(start, end, edgeStart, edgeEnd)

    if (
      intersection &&
      ((startOnEdge && arePointsClose(intersection, start)) ||
        (endOnEdge && arePointsClose(intersection, end)))
    ) {
      // The path only touches the boundary at its start or end point
      continue
    }

    if (
      intersection &&
      (arePointsClose(intersection, start) || arePointsClose(intersection, end))
    ) {
      continue
    }

    return true
  }

  return false
}

File: rust/trace-simplification/src/utils/polygon_containment.rs
use crate::types::Point2;
use crate::math_utils::{do_segments_intersect, get_segment_intersection, point_to_segment_distance};
use super::minimum_distance_between_segments::minimum_distance_between_segments;
const EPSILON: f64 = 1e-6;

fn is_point_on_segment(p: Point2, start: Point2, end: Point2) -> bool {
    point_to_segment_distance(p, start, end) <= EPSILON
}
fn are_points_close(a: Point2, b: Point2) -> bool { (a.x - b.x).abs() <= EPSILON && (a.y - b.y).abs() <= EPSILON }

pub fn is_point_in_or_on_polygon(point: Point2, polygon: &[Point2]) -> bool {
    if polygon.len() < 3 { return false; }
    for i in 0..polygon.len() {
        if is_point_on_segment(point, polygon[i], polygon[(i + 1) % polygon.len()]) { return true; }
    }
    let mut inside = false;
    let mut j = polygon.len() - 1;
    for i in 0..polygon.len() {
        let pi = polygon[i];
        let pj = polygon[j];
        let intersect = (pi.y > point.y) != (pj.y > point.y)
            && point.x < ((pj.x - pi.x) * (point.y - pi.y)) / (pj.y - pi.y) + pi.x;
        if intersect { inside = !inside; }
        j = i;
    }
    inside
}

pub fn does_segment_cross_polygon_boundary(start: Point2, end: Point2, polygon: &[Point2], margin: f64) -> bool {
    if polygon.len() < 3 { return false; }
    if !is_point_in_or_on_polygon(start, polygon) || !is_point_in_or_on_polygon(end, polygon) { return true; }
    for i in 0..polygon.len() {
        let edge_start = polygon[i];
        let edge_end = polygon[(i + 1) % polygon.len()];
        let start_on_edge = is_point_on_segment(start, edge_start, edge_end);
        let end_on_edge = is_point_on_segment(end, edge_start, edge_end);
        if start_on_edge && end_on_edge { continue; }
        if !do_segments_intersect(start, end, edge_start, edge_end) {
            if !start_on_edge && !end_on_edge
                && minimum_distance_between_segments(start, end, edge_start, edge_end) < margin - EPSILON { return true; }
            continue;
        }
        if let Some(intersection) = get_segment_intersection(start, end, edge_start, edge_end) {
            if (start_on_edge && are_points_close(intersection, start)) || (end_on_edge && are_points_close(intersection, end)) { continue; }
            if are_points_close(intersection, start) || are_points_close(intersection, end) { continue; }
        }
        return true;
    }
    false
}
